File: libs/authz-cedar/src/audit.rs
//! Audit sink for authorization decisions.
//!
//! Per ADR-0027 (and the Foundry-parity audit requirements) every
//! authorization decision must be observable: who asked for what,
//! against which resource, and what the engine answered. The
//! emission path is deliberately **fire-and-forget**: a slow audit
//! sink must NEVER stall the request hot path.
//!
//! Implementations:
//!
//! * [`NoopAuditSink`] — drops everything. Default for tests.
//! * [`TracingAuditSink`] — logs at `info!` level. Useful in dev.
//! * Production code wires a Kafka producer behind this trait that
//!   publishes to topic `audit.authz.v1`. The crate intentionally
//!   does not pull rdkafka itself — the service that owns Kafka
//!   wiring (e.g. `authorization-policy-service`) provides the impl.

use std::sync::Arc;

use async_trait::async_trait;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// Wire format for the `audit.authz.v1` Kafka topic.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AuthzAuditEvent {
    pub timestamp: DateTime<Utc>,
    pub principal: String,
    pub action: String,
    pub resource: String,
    pub decision: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tenant: Option<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub policy_ids: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub diagnostics: Vec<String>,
}

#[async_trait]
pub trait AuthzAuditSink: Send + Sync + 'static {
    async fn emit(&self, event: AuthzAuditEvent);
}

/// Drops every event. Default for tests and the in-memory engine.
#[derive(Debug, Default, Clone, Copy)]
pub struct NoopAuditSink;

#[async_trait]
impl AuthzAuditSink for NoopAuditSink {
    async fn emit(&self, _event: AuthzAuditEvent) {}
}

/// Logs every decision at `info!` level. Useful for dev / smoke runs.
#[derive(Debug, Default, Clone, Copy)]
pub struct TracingAuditSink;

#[async_trait]
impl AuthzAuditSink for TracingAuditSink {
    async fn emit(&self, event: AuthzAuditEvent) {
        tracing::info!(
            target: "authz.audit",
            principal = %event.principal,
            action = %event.action,
            resource = %event.resource,
            decision = %event.decision,
            tenant = ?event.tenant,
            policies = ?event.policy_ids,
            "authz decision"
        );
    }
}

/// Type-erased handle. The engine owns one of these.
pub type AuditSinkHandle = Arc<dyn AuthzAuditSink>;

File: libs/authz-cedar-go/audit.go
package cedarauthz

import (
	"context"
	"log/slog"
	"time"
)

// AuthzAuditEvent is the wire format for the `audit.authz.v1` Kafka
// topic. Field set + JSON tags match the Rust struct byte-for-byte so
// downstream consumers (audit-trail, audit-compliance) read events from
// either runtime indistinguishably.
type AuthzAuditEvent struct {
	Timestamp   time.Time `json:"timestamp"`
	Principal   string    `json:"principal"`
	Action      string    `json:"action"`
	Resource    string    `json:"resource"`
	Decision    string    `json:"decision"`
	Tenant      *string   `json:"tenant,omitempty"`
	PolicyIDs   []string  `json:"policy_ids,omitempty"`
	Diagnostics []string  `json:"diagnostics,omitempty"`
}

// AuthzAuditSink consumes one event at a time. Implementations MUST be
// safe for concurrent use — the engine emits from a goroutine, so
// multiple Emits may overlap.
//
// The contract is fire-and-forget: a slow sink must NEVER stall the
// request hot path. Errors are intentionally silent; sinks that need to
// surface failures should log them internally.
type AuthzAuditSink interface {
	Emit(ctx context.Context, event AuthzAuditEvent)
}

// NoopAuditSink drops every event. Default for tests and the in-memory
// engine.
type NoopAuditSink struct{}

func (NoopAuditSink) Emit(context.Context, AuthzAuditEvent) {}

// SlogAuditSink logs every decision at INFO level via the supplied
// logger (or the default logger when nil). Useful in dev / smoke runs.
//
// Replaces the Rust `TracingAuditSink` — same emission semantics, just
// using log/slog instead of the tracing crate.
type SlogAuditSink struct {
	Logger *slog.Logger
}

func (s SlogAuditSink) Emit(_ context.Context, event AuthzAuditEvent) {
	log := s.Logger
	if log == nil {
		log = slog.Default()
	}
	attrs := []any{
		slog.String("principal", event.Principal),
		slog.String("action", event.Action),
		slog.String("resource", event.Resource),
		slog.String("decision", event.Decision),
		slog.Any("policies", event.PolicyIDs),
	}
	if event.Tenant != nil {
		attrs = append(attrs, slog.String("tenant", *event.Tenant))
	}
	log.Info("authz decision", attrs...)
}
